### DISCUSSION TITLE
tobitmap extension gone from release 1.18

### DISCUSSION THREAD

**Discussion by michaelsowa** [dlemstra/Magick.NET#652] [Help]
Hi, Can someone direct me to how to accomplish the tobitmap extension that has been removed in version 7.18?

Thanks

Mike

**Comment by dlemstra** [comment]
I hope you can find the answer in the release notes: https://github.com/dlemstra/Magick.NET/releases/tag/7.18.0.0. And let me know what I can do to improve to help you find the answer.

**Reply by michaelsowa** [reply]
Hi and thank you for the reply but I can not see this change on the release notes.  Would you be able to tell me how to accomplish the tobitmap extension method with the latest version?

` Public Shared Sub AddTextWatermark(ByRef img As ImageMagick.MagickImage,
                 ByVal watermarkText As String,
                 Optional ByVal Alpha As Integer = 153,
                 Optional ByVal VAlign As WaterMarkVAlign = WaterMarkVAlign.Bottom,
                 Optional ByVal Rotate As Integer = 0,
                 Optional ByVal TextColorRGB As String = "ffffff")

            Dim bmp As Bitmap = **img.ToBitmap**
            Call ImageHandling.AddTextWatermark(bmp, watermarkText, Alpha, VAlign, Rotate, TextColorRGB)
            img = New ImageMagick.MagickImage(bmp)


        End Sub`

Thanks in advance

**Comment by dlemstra** [comment]
You will need to install the new `Magick.NET.SystemDrawing` library. It contains the `ToBitmap` method. There is also no longer a constructor that accepts a `Bitmap`. You will need to use the `MagickFactory` and call `CreateImage` to accomplish this. Or you could call the empty constructor of `var image = MagickImage()` and then `image.Read()`.

**Reply by GieltjE** [reply]
Takes some getting used to as a simple in place conversion to image.Read doesn't produce the same results without some extra tweaks. (e.g. forcing the depth to 1 bit for memorybmp (which must be forced as png or it crashes) must be done in MagickReadSettings before reading else the image is converted to some weird 3color splitted version).

**Reply by dlemstra** [reply]
The `Read(bitmap)` method should produce the same results as the `MagickImage(bitmap)` constructor. Could you explain your issue in more detail @GieltjE?

**Reply by GieltjE** [reply]
We used to generate an barcode with zxing, read it with a simple new MagickImage(<bitmap>), set the bitdepth to 1, if we do this via the .Read call we end up with a black/red/green/blue splitted image (which is fixed by specifying bitdepth 1 via the MagickReadSettings).

**Reply by dlemstra** [reply]
Can you add a small code sample to explain this?

**Reply by GieltjE** [reply]
```
            using (Bitmap result = new BarcodeWriter { Options = new AztecEncodingOptions { Width = pictureBox1.Width, Height = pictureBox1.Height } }.Write(new ZXing.QrCode.QRCodeWriter().encode("TEST", BarcodeFormat.QR_CODE, pictureBox1.Height, pictureBox1.Width)))
            {
                using (MagickImage magickImage = new MagickImage())
                {
                    using (MemoryStream memoryStream = new MemoryStream())
                    {
                        result.Save(memoryStream, ImageFormat.Png);
                        memoryStream.Position = 0;
                        magickImage.Read(memoryStream);
                        pictureBox1.Image = magickImage.ToBitmap();
                    }
                }
            }
```
A simple example

**Reply by dlemstra** [reply]
I don't have enough information to reproduce your issue locally @GieltjE. Could you create a small project on GitHub with the old code and with the new code that I can use to reproduce your issue?

**Reply by GieltjE** [reply]
![result](https://user-images.githubusercontent.com/2291132/83915270-32962600-a773-11ea-9c33-b272e4cb9a35.png)
[https://black-mail.nl/sample.7z](https://black-mail.nl/sample.7z)

**Reply by dlemstra** [reply]
Thanks for giving me the example. That made it more clear. The problem appears to be in the `ToBitmap()` method and not the `Read` method. Thanks for reporting this! Will publish a new release to fix that this weekend.

And you don't need the memory stream. You can do this:

```C#
using (MagickImage magickImage = new MagickImage())
{
    magickImage.Read(result);
    pictureBox.Image = magickImage.ToBitmap();
}
```

**Reply by GieltjE** [reply]
Good to hear it's fixable, indeed you can do .Read(Bitmap) but there's no overload for .Read(Bitmap, MagickReadSettings);

**Reply by dlemstra** [reply]
Replied to early :) Which option of the settings do you need? Not sure which one would be used in this situation. You could use the `Settings` property of the `MagickImage` instead for the setting that you want to modify.

**Reply by GieltjE** [reply]
Read(Bitmap, MagickReadSettings), updated my comment above as github purged the text.......
Using the MagickReadSettings to force it to 1bit color depth fixes the problem.

**Reply by dlemstra** [reply]
A new release that fixes your issue has been published @GieltjE. You don't need to force the bit depth anymore.
